server/src/main/java/org/apache/druid/server/coordinator/duty/CloneHistoricals.java

```diff
@@ -0,0 +1,126 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.server.coordinator.duty;
+
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.server.coordinator.CoordinatorDynamicConfig;
+import org.apache.druid.server.coordinator.DruidCoordinatorRuntimeParams;
+import org.apache.druid.server.coordinator.ServerHolder;
+import org.apache.druid.server.coordinator.loading.SegmentAction;
+import org.apache.druid.server.coordinator.loading.SegmentLoadQueueManager;
+import org.apache.druid.server.coordinator.stats.CoordinatorRunStats;
+import org.apache.druid.server.coordinator.stats.Dimension;
+import org.apache.druid.server.coordinator.stats.RowKey;
+import org.apache.druid.server.coordinator.stats.Stats;
+import org.apache.druid.timeline.DataSegment;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Handles cloning of historicals. Given the historical to historical clone mappings, based on
+ * {@link CoordinatorDynamicConfig#getCloneServers()}, copies any segments load or unload requests from the source
+ * historical to the target historical.
+ */
+public class CloneHistoricals implements CoordinatorDuty
+{
+  private static final Logger log = new Logger(CloneHistoricals.class);
+
+  @Override
+  public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+  {
+    final Map<String, String> cloneServers = params.getCoordinatorDynamicConfig().getCloneServers();
+    final CoordinatorRunStats stats = params.getCoordinatorStats();
+    final SegmentLoadQueueManager loadQueueManager = params.getLoadQueueManager();
+
+    if (cloneServers.isEmpty()) {
+      // No servers to be cloned.
+      return params;
+    }
+
+    // Create a map of host to historical.
+    final Map<String, ServerHolder> historicalMap = params.getDruidCluster()
+                                                          .getHistoricals()
+                                                          .values()
+                                                          .stream()
+                                                          .flatMap(Collection::stream)
+                                                          .collect(Collectors.toMap(
+                                                              serverHolder -> serverHolder.getServer().getHost(),
+                                                              serverHolder -> serverHolder
+                                                          ));
+
+    for (Map.Entry<String, String> entry : cloneServers.entrySet()) {
+      log.debug("Handling cloning for mapping: [%s]", entry);
+
+      final String sourceHistoricalName = entry.getKey();
+      final ServerHolder sourceServer = historicalMap.get(sourceHistoricalName);
+
+      if (sourceServer == null) {
+        log.warn(
```

Comment:
Should be error, I think, or maybe even a `log.makeAlert`.

Reply:
Changed to error